customtype/customtype.go

```diff
@@ -0,0 +1,42 @@
+package customtype
+
+import (
+	"fmt"
+	"go/ast"
+	"reflect"
+	"sync"
+)
+
+var (
+	customTypesMux sync.Mutex
+	customTypes    = make(map[reflect.Type]func(any) ast.Expr)
+)
+
+// Register registers a type that for representation in a custom manner with
+// valast. If valast encounters a value or pointer to a value of this type, it
+// will use the given render func to generate the appropriate AST representation.
+//
+// This is useful if a type's fields are private, and can only be represented
+// through a constructor - see stdtypes.go for examples.
+//
+// This mechanism currently only works with struct types.
+func Register[T any](render func(value T) ast.Expr) {
+	customTypesMux.Lock()
+	var zero T
+	t := reflect.TypeOf(zero)
+	if _, exists := customTypes[t]; exists {
+		panic(fmt.Sprintf("%T already registered", zero))
+	}
+	customTypes[t] = func(value any) ast.Expr { return render(value.(T)) }
+	customTypesMux.Unlock()
+}
+
+// Is indicates if the given reflect.Type has a custom AST representation
+// generator registered.
+func Is(rt reflect.Type) (func(any) ast.Expr, bool) {
```

Comment:
How about keeping this in the same top-level `valast` package? e.g. `valast.RegisterType`, `valast.IsType` - that seems nicer to me

Reply:
I'm not sure `valast.IsType` really works because this only applies to a super narrow subset of things right now 🤔 Additionally a subpackage keeps the scope of the access to the types map clear

But, WDYT about adding `valast.RegisterType` as an alias for `customtype.RegisterType`? I've seen this used in various libraries for exposing some internal stuff to consumers

Reply:
Just pushed a commit, PTAL!